### PR TITLE
Fixed #373 -- Support multiple column fields

### PR DESCRIPTION
**Pull request by LilyFoote**
https://code.djangoproject.com/ticket/373
Continued from #16075.

### CODE REVIEW

File: django/db/models/fields/composite.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        super().__init__(*self.cols, output_field=output_field)
+        self.alias, self.target = alias, target
+
+    def as_sql(self, compiler, connection):
+        sql, params = super().as_sql(compiler, connection)
+        return "(%s)" % sql, params
+
+    def get_lookup(self, lookup):
+        if lookup == "exact":
+            return TupleExact
```
Comment (charettes) [inline django/db/models/fields/composite.py:23]: 👋 @LilyFoote! Just passing by as [there was some recent interest in this area from the forum](https://forum.djangoproject.com/t/multiple-columns-using-mysql-in-clause/24488/7) to mention that if you're looking to implement `TupleIn` there is some pre-existing logic to support it for joins at least.

See ticket-19385, and the `ForeignObject`, `MultiColSource`, `SQLQuery.build_filter`, `RelatedIn` symbols for reference.

Comment (LilyFoote) [reply django/db/models/fields/composite.py:23]: Thanks for the pointers! Where I'm currently stuck is working out how to adapt `ForeignKey` to point to a `CompositeField`. There's something about how `ForeignKey` maps to the database column(s) that's just not clicked yet.

**Comment by Lotram** [comment]
Hey @LilyFoote , I'd really like to see this PR going forward. Is there any way I can help ? Do you know what's is still needed for this to be merged ?

**Comment by LilyFoote** [comment]
@Lotram My current sticking point is upgrading `ForeignKey` to support composite keys smoothly. Currently the `ForeignObject` parent class has some support for composite keys, but it's full of sharp edges. Smoothing those out in `ForeignKey` isn't easy though - there's a lot of inheritance and composition making it very confusing to know where the right places to change are.

I'm unlikely to make much progress over the winter, but in the spring I'll probably try to take another look.

**Comment by Lotram** [comment]
@LilyFoote Ok, I'll try to look into it too, so I can at least understand and review the changes .

**Comment by HamaBarhamou** [comment]
> @LotramMon point de friction actuel est la mise à niveau `ForeignKey`pour prendre en charge les clés composites en douceur. Actuellement, la `ForeignObject`classe parent prend en charge les clés composites, mais elle est pleine d'arêtes vives. Cependant , les lisser `ForeignKey`n'est pas facile - il y a beaucoup d'héritage et de composition, ce qui rend très déroutant de savoir où se trouvent les bons endroits pour changer.
> 
> Il est peu probable que je fasse beaucoup de progrès au cours de l'hiver, mais au printemps, j'essaierai probablement d'y jeter un autre regard.

Hi @LilyFoote ,

Thank you for sharing your current challenge with upgrading ForeignKey to support composite keys. It sounds like a complex task, especially with the intricacies of inheritance and composition in the codebase.

To assist you more effectively and refine our approach, it would be incredibly helpful if you could share any specific error codes, examples, or detailed descriptions of the issues you're encountering. Understanding the precise nature of these 'sharp edges' you mentioned will enable us to offer more targeted insights and suggestions.

If possible, any code snippets or scenarios where the current implementation falls short would be greatly valuable. This information will not only help in reproducing the issues but also in brainstorming potential solutions or workarounds.

Looking forward to diving deeper into this with you in the spring, and hoping to contribute to smoothing out these challenges in the ForeignKey upgrade.

Best regard

**Comment by sarahboyce** [comment]
Closing in favour of https://github.com/django/django/pull/18056

**Pull request closed without merging**